### ISSUE TITLE
Nodes with drawtype=allfaces and use_texture_alpha=blend not rendered properly from all angles

### ISSUE THREAD

**Issue by x2048** [minetest/minetest#11119]
##### Minetest version

```
5.4.0
```

##### OS / Hardware

Operating system: KDE Neon 5.21 / Ubuntu Focal
CPU: AMD Ryzen 7 4800H

GPU model: AMD RENOIR
OpenGL version: 4.6

##### Summary

Nodes using translucent textures with use_texture_alpha = "blend" and drawtype = "allfaces" result in visual artifacts when viewed from certain angles:
![allfaces_alpha](https://user-images.githubusercontent.com/4933697/112394408-f8474680-8cfc-11eb-9feb-e22b2ce36dc2.png)

##### Steps to reproduce

1. Use the following definition for a node:
```
{
    description = "Test",
    drawtype = "allfaces",
    tiles = {"white.png^[opacity:10"},
    inventory_image = "white.png^[opacity:129",
    use_texture_alpha = "blend",
    sunlight_propagates = true,
    walkable = false,
    pointable = false,
    diggable = false,
    floodable = true,
    paramtype = "light",
    air_equivalent = true
}
```
Where `white.png` is an 1x1 pixel all-white texture.
2. Place a cloud of such nodes (use world edit or similar mod).
3. Move around the cloud.

**Comment by sfan5** [comment]
I believe this is a duplicate of #95

**Comment by x2048** [comment]
It might be. 

I have implemented painter's algorithm in [this partial fix](https://github.com/minetest/minetest/compare/master...x2048:alpha_blend). I did not touch clouds, but the changes might fix how other transparent nodes are drawn.

What's done:
* Generate map block mesh 'towards the player'
* Order mapblocks by distance to the player
* Dynamically remesh mapblocks when the player moves

Current problems:
* Fast movement overloads the mesh generation thread. I did some minimal optimization, but it at least needs to skip mapblocks that don't have translucency.
* Moving causes artifact to appear briefly because mesh generation is asynchronous.

Advice appreciated.

**Comment by appgurueu** [comment]
My thoughts:

1. Nodes have fields telling you whether they can have semitransparency (see https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L7362-L7373). You can use this to determine whether a mapblock has semitransparency by simply looping over all nodes. That value can be cached and updated as the mapblock is updated/replaced.

2. Vertices, texture coordinates, and normals are all the same, no matter how you view a mapblock. These can be generated once and cached. What does change however is the draw order of triangles as specified by the indices. Suggestion: Recalculate just this order, or, consider generating "all" (from all look directions) possible orders and caching them (might be too memory-intensive).

3. Do the same for entities: Sort them like the mapblocks before drawing them. Use https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L4376 to determine whether an entity supports semitransparency.

4. Nontrivial: Do the same for entity & node meshes: Provide different indices for different viewing angles. Again, consider recalculating vs. caching.

5. Nontrivial: Entity-mapblock interaction: Sort all scene elements (entities & mapblocks) properly, draw them in order. If necessary, render part of a mapblock behind, part before the entity.

Alternatively:

1. Implement order-independent transparency, such as an A-buffer (extended depth buffer storing multiple fragments). Most likely won't work well with OpenGL or be quite limited and negatively impact performance.

If either of this methods is done completely, #95 and related semitransparency issues could be finally closed.

**Comment by sfan5** [comment]
> I have implemented painter's algorithm in this partial fix.

Today when I looked at a scene drawn by Minetest in RenderDoc I noticed that it draws everything back to front without using the z-buffer.
This would mean that we already use painter's algorithm and resorting the draw list is unnecessary, doesn't it?
However from a quick look at the code I can't find any sorting going on, so perhaps you can correct me on this since you've _just_ worked with the relevant code.

**Comment by x2048** [comment]
@sfan5, what I experienced is that certain viewing angles produces correct result, while other viewing angles produce garbage. This is because ClientMap::m_drawlist is an std::map<> which is ordered by std::less() by default. v3s16 implements less-than operator (see [here](https://github.com/minetest/irrlicht/blob/4931b346253c64069d45947960405c8f6afc0b55/include/vector3d.h#L87)), which orders map blocks by ascending X,Y,Z values and makes some viewing angles good and some bad. In my patch I implemented [MapBlockSorter](https://github.com/minetest/minetest/pull/11130/files#diff-b881f71a2aa67fec34166d5f482443878635bfdb588e07a987657814fb920db4R163) to order mapblocks by distance to the player which removed thick boundaries between mapblocks in all cases.
